File: Python/pi_parallel.py
from multiprocessing import Pool

num_steps = 10 ** 7
step = 1.0 / num_steps
threads = 4


def fun(part):
    pi_sum = 0.0
    for i in range(part, num_steps, threads):
        x = (i + 0.5) * step
        pi_sum += 4 / (1 + x ** 2)
    return pi_sum


if __name__ == '__main__':
    with Pool(threads) as p:
        results = p.map(fun, range(threads))
    pi = sum(results) * step

    print(pi)

File: OpenMP/pi_parallel.cpp
#include <iostream>
#include <omp.h>

const long num_steps = 1000000000;
const double step = 1.0 / (double) num_steps;
const long threads = 4

int main() {

    double pi_sum = 0.0;
    #pragma omp parallel num_threads(threads)
    {
        #pragma omp for reduction(+:pi_sum)
        for (int i = 0; i < num_steps; i++) {
            double x = (i + 0.5) * step;
            pi_sum += 4.0 / (1.0 + x * x);
        }
    }
    double pi = pi_sum * step;

    cout << pi << endl;
}